lib/matplotlib/axes/_axes.py

```diff
@@ -2745,7 +2747,12 @@ def sign(x):
             if np.isnan(dat):
                 lbl = ''
 
-            annotation = self.annotate(fmt % value if lbl is None else lbl,
+            if lbl is None:
+                try:
+                    lbl = fmt(value)
+                except TypeError:
+                    lbl = cbook._auto_format_str(fmt, value)
```

Comment:
```suggestion
                if isinstance(fmt, str):
                    lbl = cbook._auto_format_str(fmt, value)
                elif callable(fmt):
                    lbl = fmt(value)
                else:
                    raise TypeError("fmt must be a str or callable")
```
In this case we should do explicit checks:

- If `fmt()` is callable but produces a TypeError internally, the try-except version would fall back to the str code, which would not work and create a strange error message.
- With only two additional lines, we can add a proper error message, if the input type is unsupported.

Side-note: I have put the str-test first because that's the most-common (and default) usage. The code is more readable when you go in the order common-less_used-error in if chains.